baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/StreamToIterator.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import java.util.stream.Stream;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.WARNING,
+        summary =
+                "Avoid calling iterator on stream, as it buffers the stream into memory and may cause unexpected OOMs."
+                        + "For small streams, consider collecting the stream first to make it clear to readers that"
+                        + " the stream will be loaded into memory.")
```

Comment:
This is fair, and a subtle interaction that is probably not expected in most cases, but I'm a bit worried about how common it may be. The suggested alternative isn't necessarily better, but makes it more obvious reading the code that data may be buffered -- in many cases, the stream isn't fully buffered, so the suggestion would result in adding buffering (albeit at a known point, with fewer spooky interactions).

In general when we are worried about resource utilization (cpu and/or memory) we should stay clear of streams imo.

Reply:
That's true - I was concerned that people probably use streams often for innocuous use-cases where buffering really won't make a difference. Our team has been bit by this twice, but maybe we really should just not be using streams at all.